### PR TITLE
Materialization status in graphviz

### PR DESCRIPTION
**Pull request by ekmartin**
The snapshotting implementation needs to know which domains have at least one materialized node, but I thought it might be useful to show it in Blender's graphviz output as well.

I'm not sure if the way I'm retrieving materialization status for a node is correct, so let me know if there's a better way to achieve the same. 

This is what it outputs for the example:
image

Feel free to bikeshed on how it should be shown in the output - currently it uses a half circle  for partial and a full circle for full materializations.

### CODE REVIEW

**Comment by ms705** [comment]
NICE! I'm a fan :+1:

File: dataflow/src/node/debug.rs [hunk 1]
```diff
@@ -40,33 +50,42 @@ impl Node {
...
-            NodeType::Egress { .. } => s.push_str(&format!("{{ {} | (egress) }}", addr)),
+            NodeType::Ingress => s.push_str(&format!(
+                "{{ {{ {} {} }} | (ingress) }}",
+                addr,
+                materialized
+            )),
+            NodeType::Egress { .. } => {
+                s.push_str(&format!("{{ {{ {} {} }} | (egress) }}", addr, materialized))
+            }
             NodeType::Sharder { .. } => s.push_str(&format!("{{ {} | (sharder) }}", addr)),
```
Comment (ekmartin) [inline dataflow/src/node/debug.rs]: Can a `Sharder` be materialized?

Comment (jonhoo) [reply dataflow/src/node/debug.rs]: No, a `Sharder` cannot be materialized afaik.

File: dataflow/src/node/debug.rs [hunk 2]
```diff
@@ -40,33 +50,42 @@ impl Node {
...
             NodeType::Source => s.push_str("(source)"),
             NodeType::Dropped => s.push_str("✗"),
-            NodeType::Ingress => s.push_str(&format!("{{ {} | (ingress) }}", addr)),
-            NodeType::Egress { .. } => s.push_str(&format!("{{ {} | (egress) }}", addr)),
+            NodeType::Ingress => s.push_str(&format!(
+                "{{ {{ {} {} }} | (ingress) }}",
+                addr,
+                materialized
+            )),
+            NodeType::Egress { .. } => {
```
Comment (ms705) [inline dataflow/src/node/debug.rs]: I don't think an `Egress` node would currently ever be materialized (@jonhoo -- ?), though this code is obviously still correct in that case.

File: src/controller/migrate/materialization/mod.rs [hunk 3]
```diff
@@ -404,6 +404,21 @@ impl Materializations {
         assert!(replay_obligations.is_empty());
     }
 
+    /// Retrieves the materialization status of a given node, or None
+    /// if the node isn't materialized.
+    pub fn get_status(&self, index: &NodeIndex, node: &Node) -> MaterializationStatus {
+        let is_materialized = self.have.contains_key(index)
+            || node.with_reader(|r| r.is_materialized()).unwrap_or(false);
```
Comment (ms705) [inline src/controller/migrate/materialization/mod.rs]: Correct, but aren't readers always materialized? (`is_materialized`returns a bool, so it doesn't differentiate full/partial either)

File: dataflow/src/node/debug.rs (hunk 2 above)
Comment (jonhoo) [reply dataflow/src/node/debug.rs]: Yeah, `Egress` is never materialized.

File: dataflow/src/node/debug.rs [hunk 4]
```diff
@@ -40,10 +50,15 @@ impl Node {
...
         match self.inner {
             NodeType::Source => s.push_str("(source)"),
             NodeType::Dropped => s.push_str("✗"),
-            NodeType::Ingress => s.push_str(&format!("{{ {} | (ingress) }}", addr)),
+            NodeType::Ingress => s.push_str(&format!(
+                "{{ {{ {} {} }} | (ingress) }}",
+                addr,
+                materialized
+            )),
             NodeType::Egress { .. } => s.push_str(&format!("{{ {} | (egress) }}", addr)),
```
Comment (ekmartin) [inline dataflow/src/node/debug.rs]: Removed `materialization` from `Egress`

**Comment by ms705** [comment]
LGTM :+1: Let's merge?

**Pull request merged**